runtime/src/main/java/org/corfudb/runtime/collections/TxnContext.java

```diff
@@ -0,0 +1,325 @@
+     * @param <K>       Type of Key.
+     * @param <V>       Type of Value.
+     * @param <I>       Type of index/secondary key.
+     * @return Result of the query.
+     */
+    @Nonnull
+    public <K extends Message, V extends Message, M extends Message, I extends Comparable<I>>
+    List<CorfuStoreEntry<K, V, M>> getByIndex(@Nonnull Table<K, V, M> table,
+                                              @Nonnull final String indexName,
+                                              @Nonnull final I indexKey) {
+        txBeginMaybe();
+        operations.forEach(Runnable::run); // Apply all pending mutations for reads to see it.
+        operations.clear();
+        return table.getByIndex(indexName, indexKey);
+    }
+
+    /**
+     * Gets the count of records in the table at a particular timestamp.
+     *
+     * @param table - the table whose count is requested.
+     * @return Count of records.
+     */
+    public <K extends Message, V extends Message, M extends Message>
+    int count(@Nonnull final Table<K, V, M> table) {
+        txBeginMaybe();
+        operations.forEach(Runnable::run); // Apply all pending mutations for reads to see it.
+        operations.clear();
+        return table.count();
+    }
+
+    /**
+     * Gets all the keys of a table.
+     *
+     * @param table - the table whose keys are requested.
+     * @return keyset of the table
+     */
+    public <K extends Message, V extends Message, M extends Message>
+    Set<K> keySet(@Nonnull final Table<K, V, M> table) {
+        txBeginMaybe();
+        operations.forEach(Runnable::run); // Apply all pending mutations for reads to see it.
+        operations.clear();
+        return table.keySet();
+    }
+
+    /**
+     * Scan and filter by entry.
+     *
+     * @param entryPredicate Predicate to filter the entries.
+     * @return Collection of filtered entries.
+     */
+    public <K extends Message, V extends Message, M extends Message>
+    List<CorfuStoreEntry<K, V, M>> executeQuery(@Nonnull final Table<K, V, M> table,
+                                                @Nonnull final Predicate<CorfuStoreEntry<K, V, M>> entryPredicate) {
+        txBeginMaybe();
+        operations.forEach(Runnable::run); // Apply all pending mutations for reads to see it.
+        operations.clear();
+        return table.scanAndFilterByEntry(entryPredicate);
+    }
+
+    /**
+     * Start the corfu transaction. When not reading transactional writes, the internal
+     * corfu transaction will begin on transaction commit()
+     */
+    private boolean txBeginMaybe() {
+        if (TransactionalContext.isInTransaction()) {
+            return true;
+        }
+        Transaction.TransactionBuilder transactionBuilder = this.objectsView
+                .TXBuild()
+                .type(TransactionType.WRITE_AFTER_WRITE);
+        if (isolationLevel.getTimestamp() != Token.UNINITIALIZED) {
+            transactionBuilder.snapshot(isolationLevel.getTimestamp());
+        }
```

Comment:
if the timestamp has not been specified, we need to take the snapshot at the current timestamp, right?  Something like transactionBuilder.snapshot(currentTimestamp)..